pkg/apis/core/v1/helper/helpers.go

```diff
@@ -304,7 +304,11 @@ func MatchNodeSelectorTerms(
 	nodeLabels labels.Set,
 	nodeFields fields.Set,
 ) bool {
-	for _, req := range nodeSelectorTerms {
+	for _, r := range nodeSelectorTerms {
+		// if we don't make a copy
+		// in NodeSelectorRequirementsAsSelector(), NewRequirement() will
+		// modify input values via sort.Strings(vals)
+		req := r.DeepCopy()
```

Comment:
Good catch ! 

IMO, it's better to enhance `NodeSelectorRequirementsAsSelector` to avoid touching the parameters. @liggitt , WDYT?

Reply:
Agree, good catch. I do think we want to avoid a full deep copy like this if we can… this is evaluated frequently and could be a significant allocation hotspot. Can we remove the need for mutation or just the string slice before sort?

Reply:
@k82cn maybe we can walk one step further - make changes in `NewRequirement()` (in selector.go) as there are other invocations like TopologySelectorRequirementsAsSelector().

 @liggitt make a copy of string slice before sort (`NewRequirement()` in selector.go) might be a safe solution. I'm not confident at removing the whole sort thing... (suppose this is what you mean by "remove the need for mutation")

Reply:
Good catch !
I’d be more in favor of just removing the sort.

Reply:
>  just the string slice before sort?

Both value & key are sorted, any background for that?  There's a comment that `// sort to have deterministic string representation`, do we need to keep that?



https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/labels/selector.go#L851